// element-plus/lib/el-dialog/index.js
import { ref, computed, watch, nextTick, onMounted, defineComponent, resolveComponent, resolveDirective, openBlock, createBlock, Fragment, Teleport, createVNode, Transition, withCtx, withDirectives, renderSlot, toDisplayString, createCommentVNode } from 'vue';
import { TrapFocus } from '../directives';
import { isValidWidthUnit } from '../utils/validators';
import { Overlay } from '../el-overlay';
import isServer from '../utils/isServer';
import { UPDATE_MODEL_EVENT } from '../utils/constants';
import PopupManager from '../utils/popup-manager';
import { clearTimer } from '../utils/util';
import { useLockScreen, useModal, useRestoreActive } from '../hooks';

const CLOSE_EVENT = 'close';
const OPEN_EVENT = 'open';
const CLOSED_EVENT = 'closed';
const OPENED_EVENT = 'opened';
function useDialog (props, ctx) {
    const visible = ref(false);
    const closed = ref(false);
    const dialogRef = ref(null);
    const openTimer = ref(null);
    const closeTimer = ref(null);
    const zIndex = ref(props.zIndex || PopupManager.nextZIndex());
    const modalRef = ref(null);
    const style = computed(() => {
        const style = {};
        if (!props.fullscreen) {
            style.marginTop = props.top;
            if (props.width) {
                style.width = props.width;
            }
        }
        return style;
    });
    function afterEnter() {
        ctx.emit(OPENED_EVENT);
    }
    function afterLeave() {
        ctx.emit(CLOSED_EVENT);
        ctx.emit(UPDATE_MODEL_EVENT, false);
    }
    function open() {
        clearTimer(closeTimer);
        clearTimer(openTimer);
        if (props.openDelay && props.openDelay > 0) {
            openTimer.value = window.setTimeout(() => {
                openTimer.value = null;
                doOpen();
            }, props.openDelay);
        }
        else {
            doOpen();
        }
    }
    function close() {
        clearTimer(openTimer);
        clearTimer(closeTimer);
        if (props.closeDelay && props.closeDelay > 0) {
            closeTimer.value = window.setTimeout(() => {
                closeTimer.value = null;
                doClose();
            }, props.closeDelay);
        }
        else {
            doClose();
        }
    }
    function hide(shouldCancel) {
        if (shouldCancel)
            return;
        closed.value = true;
        visible.value = false;
    }
    function handleClose() {
        if (props.beforeClose) {
            props.beforeClose(hide);
        }
        else {
            close();
        }
    }
    function onModalClick() {
        if (props.closeOnClickModal) {
            handleClose();
        }
    }
    function doOpen() {
        if (isServer) {
            return;
        }
        visible.value = true;
    }
    function doClose() {
        visible.value = false;
    }
    if (props.lockScroll) {
        useLockScreen(visible);
    }
    if (props.closeOnPressEscape) {
        useModal({
            handleClose,
        }, visible);
    }
    useRestoreActive(visible);
    watch(() => props.modelValue, val => {
        if (val) {
            closed.value = false;
            open();
            ctx.emit(OPEN_EVENT);
            nextTick(() => {
                if (dialogRef.value) {
                    dialogRef.value.scrollTop = 0;
                }
            });
        }
        else {
            close();
            if (!closed.value) {
                ctx.emit(CLOSE_EVENT);
            }
        }
    });
    onMounted(() => {
        if (props.modelValue) {
            visible.value = true;
            open();
        }
    });
    return {
        afterEnter,
        afterLeave,
        handleClose,
        onModalClick,
        closed,
        dialogRef,
        style,
        modalRef,
        visible,
        zIndex,
    };
}

var script = defineComponent({
    name: 'ElDialog',
    components: {
        'el-overlay': Overlay,
    },
    directives: {
        TrapFocus,
    },
    props: {
        appendToBody: {
            type: Boolean,
            default: false,
        },
        beforeClose: {
            type: Function,
        },
        destroyOnClose: {
            type: Boolean,
            default: false,
        },
        center: {
            type: Boolean,
            default: false,
        },
        customClass: {
            type: String,
            default: '',
        },
        closeOnClickModal: {
            type: Boolean,
            default: true,
        },
        closeOnPressEscape: {
            type: Boolean,
            default: true,
        },
        fullscreen: {
            type: Boolean,
            default: false,
        },
        lockScroll: {
            type: Boolean,
            default: true,
        },
        modal: {
            type: Boolean,
            default: true,
        },
        showClose: {
            type: Boolean,
            default: true,
        },
        title: {
            type: String,
            default: '',
        },
        openDelay: {
            type: Number,
            default: 0,
        },
        closeDelay: {
            type: Number,
            default: 0,
        },
        top: {
            type: String,
            default: '15vh',
        },
        modelValue: {
            type: Boolean,
            required: true,
        },
        width: {
            type: String,
            default: '50%',
            validator: isValidWidthUnit,
        },
        zIndex: {
            type: Number,
        },
    },
    emits: [
        OPEN_EVENT,
        OPENED_EVENT,
        CLOSE_EVENT,
        CLOSED_EVENT,
        UPDATE_MODEL_EVENT,
    ],
    setup(props, ctx) {
        return useDialog(props, ctx);
    },
});

const _hoisted_1 = { class: "el-dialog__header" };
const _hoisted_2 = { class: "el-dialog__title" };
const _hoisted_3 = /*#__PURE__*/createVNode("i", { class: "el-dialog__close el-icon el-icon-close" }, null, -1 /* HOISTED */);
const _hoisted_4 = { class: "el-dialog__body" };
const _hoisted_5 = {
  key: 0,
  class: "el-dialog__footer"
};

function render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_el_overlay = resolveComponent("el-overlay");
  const _directive_trap_focus = resolveDirective("trap-focus");

  return (_ctx.destroyOnClose && !_ctx.visible)
    ? (openBlock(), createBlock(Fragment, { key: 0 }, [], 64 /* STABLE_FRAGMENT */))
    : (openBlock(), createBlock(Teleport, {
        key: 1,
        to: "body",
        disabled: !_ctx.appendToBody
      }, [
        createVNode(Transition, {
          name: "dialog-fade",
          onAfterEnter: _ctx.afterEnter,
          onAfterLeave: _ctx.afterLeave
        }, {
          default: withCtx(() => [
            (_ctx.visible)
              ? (openBlock(), createBlock(_component_el_overlay, {
                  key: 0,
                  "z-index": _ctx.zIndex,
                  mask: _ctx.modal,
                  onClick: _ctx.onModalClick
                }, {
                  default: withCtx(() => [
                    withDirectives(createVNode("div", {
                      ref: "dialogRef",
                      class: [
              'el-dialog',
              {
                'is-fullscreen': _ctx.fullscreen,
                'el-dialog--center': _ctx.center,
              },
              _ctx.customClass,
            ],
                      "aria-modal": "true",
                      role: "dialog",
                      "aria-label": _ctx.title || 'dialog',
                      style: _ctx.style,
                      onClick: _cache[2] || (_cache[2] = $event => ($event.stopPropagation()))
                    }, [
                      createVNode("div", _hoisted_1, [
                        renderSlot(_ctx.$slots, "header", {}, () => [
                          createVNode("span", _hoisted_2, toDisplayString(_ctx.title), 1 /* TEXT */)
                        ]),
                        createVNode("button", {
                          "aria-label": "close",
                          class: "el-dialog__headerbtn",
                          type: "button",
                          onClick: _cache[1] || (_cache[1] = (...args) => (_ctx.handleClose(...args)))
                        }, [
                          _hoisted_3
                        ])
                      ]),
                      createVNode("div", _hoisted_4, [
                        renderSlot(_ctx.$slots, "default")
                      ]),
                      (_ctx.$slots.footer)
                        ? (openBlock(), createBlock("div", _hoisted_5, [
                            renderSlot(_ctx.$slots, "footer")
                          ]))
                        : createCommentVNode("v-if", true)
                    ], 14 /* CLASS, STYLE, PROPS */, ["aria-label"]), [
                      [_directive_trap_focus]
                    ])
                  ]),
                  _: 3
                }, 8 /* PROPS */, ["z-index", "mask", "onClick"]))
              : createCommentVNode("v-if", true)
          ]),
          _: 1
        }, 8 /* PROPS */, ["onAfterEnter", "onAfterLeave"])
      ], 8 /* PROPS */, ["disabled"]))
}

script.render = render;
script.__file = "packages/dialog/src/index.vue";

script.install = (app) => {
    app.component(script.name, script);
};

export default script;
