// vant/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.install = install;
Object.defineProperty(exports, "ActionBar", {
  enumerable: true,
  get: function get() {
    return _actionBar.default;
  }
});
Object.defineProperty(exports, "ActionBarButton", {
  enumerable: true,
  get: function get() {
    return _actionBarButton.default;
  }
});
Object.defineProperty(exports, "ActionBarIcon", {
  enumerable: true,
  get: function get() {
    return _actionBarIcon.default;
  }
});
Object.defineProperty(exports, "ActionSheet", {
  enumerable: true,
  get: function get() {
    return _actionSheet.default;
  }
});
Object.defineProperty(exports, "AddressEdit", {
  enumerable: true,
  get: function get() {
    return _addressEdit.default;
  }
});
Object.defineProperty(exports, "AddressList", {
  enumerable: true,
  get: function get() {
    return _addressList.default;
  }
});
Object.defineProperty(exports, "Area", {
  enumerable: true,
  get: function get() {
    return _area.default;
  }
});
Object.defineProperty(exports, "Badge", {
  enumerable: true,
  get: function get() {
    return _badge.default;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function get() {
    return _button.default;
  }
});
Object.defineProperty(exports, "Calendar", {
  enumerable: true,
  get: function get() {
    return _calendar.default;
  }
});
Object.defineProperty(exports, "Card", {
  enumerable: true,
  get: function get() {
    return _card.default;
  }
});
Object.defineProperty(exports, "Cell", {
  enumerable: true,
  get: function get() {
    return _cell.default;
  }
});
Object.defineProperty(exports, "CellGroup", {
  enumerable: true,
  get: function get() {
    return _cellGroup.default;
  }
});
Object.defineProperty(exports, "Checkbox", {
  enumerable: true,
  get: function get() {
    return _checkbox.default;
  }
});
Object.defineProperty(exports, "CheckboxGroup", {
  enumerable: true,
  get: function get() {
    return _checkboxGroup.default;
  }
});
Object.defineProperty(exports, "Circle", {
  enumerable: true,
  get: function get() {
    return _circle.default;
  }
});
Object.defineProperty(exports, "Col", {
  enumerable: true,
  get: function get() {
    return _col.default;
  }
});
Object.defineProperty(exports, "Collapse", {
  enumerable: true,
  get: function get() {
    return _collapse.default;
  }
});
Object.defineProperty(exports, "CollapseItem", {
  enumerable: true,
  get: function get() {
    return _collapseItem.default;
  }
});
Object.defineProperty(exports, "ContactCard", {
  enumerable: true,
  get: function get() {
    return _contactCard.default;
  }
});
Object.defineProperty(exports, "ContactEdit", {
  enumerable: true,
  get: function get() {
    return _contactEdit.default;
  }
});
Object.defineProperty(exports, "ContactList", {
  enumerable: true,
  get: function get() {
    return _contactList.default;
  }
});
Object.defineProperty(exports, "CountDown", {
  enumerable: true,
  get: function get() {
    return _countDown.default;
  }
});
Object.defineProperty(exports, "Coupon", {
  enumerable: true,
  get: function get() {
    return _coupon.default;
  }
});
Object.defineProperty(exports, "CouponCell", {
  enumerable: true,
  get: function get() {
    return _couponCell.default;
  }
});
Object.defineProperty(exports, "CouponList", {
  enumerable: true,
  get: function get() {
    return _couponList.default;
  }
});
Object.defineProperty(exports, "DatetimePicker", {
  enumerable: true,
  get: function get() {
    return _datetimePicker.default;
  }
});
Object.defineProperty(exports, "Dialog", {
  enumerable: true,
  get: function get() {
    return _dialog.default;
  }
});
Object.defineProperty(exports, "Divider", {
  enumerable: true,
  get: function get() {
    return _divider.default;
  }
});
Object.defineProperty(exports, "DropdownItem", {
  enumerable: true,
  get: function get() {
    return _dropdownItem.default;
  }
});
Object.defineProperty(exports, "DropdownMenu", {
  enumerable: true,
  get: function get() {
    return _dropdownMenu.default;
  }
});
Object.defineProperty(exports, "Empty", {
  enumerable: true,
  get: function get() {
    return _empty.default;
  }
});
Object.defineProperty(exports, "Field", {
  enumerable: true,
  get: function get() {
    return _field.default;
  }
});
Object.defineProperty(exports, "Form", {
  enumerable: true,
  get: function get() {
    return _form.default;
  }
});
Object.defineProperty(exports, "Grid", {
  enumerable: true,
  get: function get() {
    return _grid.default;
  }
});
Object.defineProperty(exports, "GridItem", {
  enumerable: true,
  get: function get() {
    return _gridItem.default;
  }
});
Object.defineProperty(exports, "Icon", {
  enumerable: true,
  get: function get() {
    return _icon.default;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function get() {
    return _image.default;
  }
});
Object.defineProperty(exports, "ImagePreview", {
  enumerable: true,
  get: function get() {
    return _imagePreview.default;
  }
});
Object.defineProperty(exports, "IndexAnchor", {
  enumerable: true,
  get: function get() {
    return _indexAnchor.default;
  }
});
Object.defineProperty(exports, "IndexBar", {
  enumerable: true,
  get: function get() {
    return _indexBar.default;
  }
});
Object.defineProperty(exports, "Lazyload", {
  enumerable: true,
  get: function get() {
    return _lazyload.default;
  }
});
Object.defineProperty(exports, "List", {
  enumerable: true,
  get: function get() {
    return _list.default;
  }
});
Object.defineProperty(exports, "Loading", {
  enumerable: true,
  get: function get() {
    return _loading.default;
  }
});
Object.defineProperty(exports, "Locale", {
  enumerable: true,
  get: function get() {
    return _locale.default;
  }
});
Object.defineProperty(exports, "NavBar", {
  enumerable: true,
  get: function get() {
    return _navBar.default;
  }
});
Object.defineProperty(exports, "NoticeBar", {
  enumerable: true,
  get: function get() {
    return _noticeBar.default;
  }
});
Object.defineProperty(exports, "Notify", {
  enumerable: true,
  get: function get() {
    return _notify.default;
  }
});
Object.defineProperty(exports, "NumberKeyboard", {
  enumerable: true,
  get: function get() {
    return _numberKeyboard.default;
  }
});
Object.defineProperty(exports, "Overlay", {
  enumerable: true,
  get: function get() {
    return _overlay.default;
  }
});
Object.defineProperty(exports, "Pagination", {
  enumerable: true,
  get: function get() {
    return _pagination.default;
  }
});
Object.defineProperty(exports, "PasswordInput", {
  enumerable: true,
  get: function get() {
    return _passwordInput.default;
  }
});
Object.defineProperty(exports, "Picker", {
  enumerable: true,
  get: function get() {
    return _picker.default;
  }
});
Object.defineProperty(exports, "Popover", {
  enumerable: true,
  get: function get() {
    return _popover.default;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function get() {
    return _popup.default;
  }
});
Object.defineProperty(exports, "Progress", {
  enumerable: true,
  get: function get() {
    return _progress.default;
  }
});
Object.defineProperty(exports, "PullRefresh", {
  enumerable: true,
  get: function get() {
    return _pullRefresh.default;
  }
});
Object.defineProperty(exports, "Radio", {
  enumerable: true,
  get: function get() {
    return _radio.default;
  }
});
Object.defineProperty(exports, "RadioGroup", {
  enumerable: true,
  get: function get() {
    return _radioGroup.default;
  }
});
Object.defineProperty(exports, "Rate", {
  enumerable: true,
  get: function get() {
    return _rate.default;
  }
});
Object.defineProperty(exports, "Row", {
  enumerable: true,
  get: function get() {
    return _row.default;
  }
});
Object.defineProperty(exports, "Search", {
  enumerable: true,
  get: function get() {
    return _search.default;
  }
});
Object.defineProperty(exports, "ShareSheet", {
  enumerable: true,
  get: function get() {
    return _shareSheet.default;
  }
});
Object.defineProperty(exports, "Sidebar", {
  enumerable: true,
  get: function get() {
    return _sidebar.default;
  }
});
Object.defineProperty(exports, "SidebarItem", {
  enumerable: true,
  get: function get() {
    return _sidebarItem.default;
  }
});
Object.defineProperty(exports, "Skeleton", {
  enumerable: true,
  get: function get() {
    return _skeleton.default;
  }
});
Object.defineProperty(exports, "Slider", {
  enumerable: true,
  get: function get() {
    return _slider.default;
  }
});
Object.defineProperty(exports, "Step", {
  enumerable: true,
  get: function get() {
    return _step.default;
  }
});
Object.defineProperty(exports, "Stepper", {
  enumerable: true,
  get: function get() {
    return _stepper.default;
  }
});
Object.defineProperty(exports, "Steps", {
  enumerable: true,
  get: function get() {
    return _steps.default;
  }
});
Object.defineProperty(exports, "Sticky", {
  enumerable: true,
  get: function get() {
    return _sticky.default;
  }
});
Object.defineProperty(exports, "SubmitBar", {
  enumerable: true,
  get: function get() {
    return _submitBar.default;
  }
});
Object.defineProperty(exports, "Swipe", {
  enumerable: true,
  get: function get() {
    return _swipe.default;
  }
});
Object.defineProperty(exports, "SwipeCell", {
  enumerable: true,
  get: function get() {
    return _swipeCell.default;
  }
});
Object.defineProperty(exports, "SwipeItem", {
  enumerable: true,
  get: function get() {
    return _swipeItem.default;
  }
});
Object.defineProperty(exports, "Switch", {
  enumerable: true,
  get: function get() {
    return _switch.default;
  }
});
Object.defineProperty(exports, "Tab", {
  enumerable: true,
  get: function get() {
    return _tab.default;
  }
});
Object.defineProperty(exports, "Tabbar", {
  enumerable: true,
  get: function get() {
    return _tabbar.default;
  }
});
Object.defineProperty(exports, "TabbarItem", {
  enumerable: true,
  get: function get() {
    return _tabbarItem.default;
  }
});
Object.defineProperty(exports, "Tabs", {
  enumerable: true,
  get: function get() {
    return _tabs.default;
  }
});
Object.defineProperty(exports, "Tag", {
  enumerable: true,
  get: function get() {
    return _tag.default;
  }
});
Object.defineProperty(exports, "Toast", {
  enumerable: true,
  get: function get() {
    return _toast.default;
  }
});
Object.defineProperty(exports, "TreeSelect", {
  enumerable: true,
  get: function get() {
    return _treeSelect.default;
  }
});
Object.defineProperty(exports, "Uploader", {
  enumerable: true,
  get: function get() {
    return _uploader.default;
  }
});
exports.default = exports.version = void 0;

var _actionBar = _interopRequireDefault(require("./action-bar"));

var _actionBarButton = _interopRequireDefault(require("./action-bar-button"));

var _actionBarIcon = _interopRequireDefault(require("./action-bar-icon"));

var _actionSheet = _interopRequireDefault(require("./action-sheet"));

var _addressEdit = _interopRequireDefault(require("./address-edit"));

var _addressList = _interopRequireDefault(require("./address-list"));

var _area = _interopRequireDefault(require("./area"));

var _badge = _interopRequireDefault(require("./badge"));

var _button = _interopRequireDefault(require("./button"));

var _calendar = _interopRequireDefault(require("./calendar"));

var _card = _interopRequireDefault(require("./card"));

var _cell = _interopRequireDefault(require("./cell"));

var _cellGroup = _interopRequireDefault(require("./cell-group"));

var _checkbox = _interopRequireDefault(require("./checkbox"));

var _checkboxGroup = _interopRequireDefault(require("./checkbox-group"));

var _circle = _interopRequireDefault(require("./circle"));

var _col = _interopRequireDefault(require("./col"));

var _collapse = _interopRequireDefault(require("./collapse"));

var _collapseItem = _interopRequireDefault(require("./collapse-item"));

var _contactCard = _interopRequireDefault(require("./contact-card"));

var _contactEdit = _interopRequireDefault(require("./contact-edit"));

var _contactList = _interopRequireDefault(require("./contact-list"));

var _countDown = _interopRequireDefault(require("./count-down"));

var _coupon = _interopRequireDefault(require("./coupon"));

var _couponCell = _interopRequireDefault(require("./coupon-cell"));

var _couponList = _interopRequireDefault(require("./coupon-list"));

var _datetimePicker = _interopRequireDefault(require("./datetime-picker"));

var _dialog = _interopRequireDefault(require("./dialog"));

var _divider = _interopRequireDefault(require("./divider"));

var _dropdownItem = _interopRequireDefault(require("./dropdown-item"));

var _dropdownMenu = _interopRequireDefault(require("./dropdown-menu"));

var _empty = _interopRequireDefault(require("./empty"));

var _field = _interopRequireDefault(require("./field"));

var _form = _interopRequireDefault(require("./form"));

var _grid = _interopRequireDefault(require("./grid"));

var _gridItem = _interopRequireDefault(require("./grid-item"));

var _icon = _interopRequireDefault(require("./icon"));

var _image = _interopRequireDefault(require("./image"));

var _imagePreview = _interopRequireDefault(require("./image-preview"));

var _indexAnchor = _interopRequireDefault(require("./index-anchor"));

var _indexBar = _interopRequireDefault(require("./index-bar"));

var _lazyload = _interopRequireDefault(require("./lazyload"));

var _list = _interopRequireDefault(require("./list"));

var _loading = _interopRequireDefault(require("./loading"));

var _locale = _interopRequireDefault(require("./locale"));

var _navBar = _interopRequireDefault(require("./nav-bar"));

var _noticeBar = _interopRequireDefault(require("./notice-bar"));

var _notify = _interopRequireDefault(require("./notify"));

var _numberKeyboard = _interopRequireDefault(require("./number-keyboard"));

var _overlay = _interopRequireDefault(require("./overlay"));

var _pagination = _interopRequireDefault(require("./pagination"));

var _passwordInput = _interopRequireDefault(require("./password-input"));

var _picker = _interopRequireDefault(require("./picker"));

var _popover = _interopRequireDefault(require("./popover"));

var _popup = _interopRequireDefault(require("./popup"));

var _progress = _interopRequireDefault(require("./progress"));

var _pullRefresh = _interopRequireDefault(require("./pull-refresh"));

var _radio = _interopRequireDefault(require("./radio"));

var _radioGroup = _interopRequireDefault(require("./radio-group"));

var _rate = _interopRequireDefault(require("./rate"));

var _row = _interopRequireDefault(require("./row"));

var _search = _interopRequireDefault(require("./search"));

var _shareSheet = _interopRequireDefault(require("./share-sheet"));

var _sidebar = _interopRequireDefault(require("./sidebar"));

var _sidebarItem = _interopRequireDefault(require("./sidebar-item"));

var _skeleton = _interopRequireDefault(require("./skeleton"));

var _slider = _interopRequireDefault(require("./slider"));

var _step = _interopRequireDefault(require("./step"));

var _stepper = _interopRequireDefault(require("./stepper"));

var _steps = _interopRequireDefault(require("./steps"));

var _sticky = _interopRequireDefault(require("./sticky"));

var _submitBar = _interopRequireDefault(require("./submit-bar"));

var _swipe = _interopRequireDefault(require("./swipe"));

var _swipeCell = _interopRequireDefault(require("./swipe-cell"));

var _swipeItem = _interopRequireDefault(require("./swipe-item"));

var _switch = _interopRequireDefault(require("./switch"));

var _tab = _interopRequireDefault(require("./tab"));

var _tabbar = _interopRequireDefault(require("./tabbar"));

var _tabbarItem = _interopRequireDefault(require("./tabbar-item"));

var _tabs = _interopRequireDefault(require("./tabs"));

var _tag = _interopRequireDefault(require("./tag"));

var _toast = _interopRequireDefault(require("./toast"));

var _treeSelect = _interopRequireDefault(require("./tree-select"));

var _uploader = _interopRequireDefault(require("./uploader"));

var version = '3.0.0-beta.10';
exports.version = version;

function install(app) {
  var components = [_actionBar.default, _actionBarButton.default, _actionBarIcon.default, _actionSheet.default, _addressEdit.default, _addressList.default, _area.default, _badge.default, _button.default, _calendar.default, _card.default, _cell.default, _cellGroup.default, _checkbox.default, _checkboxGroup.default, _circle.default, _col.default, _collapse.default, _collapseItem.default, _contactCard.default, _contactEdit.default, _contactList.default, _countDown.default, _coupon.default, _couponCell.default, _couponList.default, _datetimePicker.default, _dialog.default, _divider.default, _dropdownItem.default, _dropdownMenu.default, _empty.default, _field.default, _form.default, _grid.default, _gridItem.default, _icon.default, _image.default, _imagePreview.default, _indexAnchor.default, _indexBar.default, _list.default, _loading.default, _locale.default, _navBar.default, _noticeBar.default, _notify.default, _numberKeyboard.default, _overlay.default, _pagination.default, _passwordInput.default, _picker.default, _popover.default, _popup.default, _progress.default, _pullRefresh.default, _radio.default, _radioGroup.default, _rate.default, _row.default, _search.default, _shareSheet.default, _sidebar.default, _sidebarItem.default, _skeleton.default, _slider.default, _step.default, _stepper.default, _steps.default, _sticky.default, _submitBar.default, _swipe.default, _swipeCell.default, _swipeItem.default, _switch.default, _tab.default, _tabbar.default, _tabbarItem.default, _tabs.default, _tag.default, _toast.default, _treeSelect.default, _uploader.default];
  components.forEach(function (item) {
    if (item.install) {
      app.use(item);
    } else if (item.name) {
      app.component(item.name, item);
    }
  });
}

var _default = {
  install: install,
  version: version
};
exports.default = _default;
